.github/workflows/build-and-test.yml

```diff
@@ -1,91 +1,59 @@
       matrix:
         os:
           - macos-latest
           - ubuntu-latest
           - windows-latest
+        ocaml-compiler:
+          - 4.12.x
 
-        ocaml-version:
-          - 4.12.0
+    runs-on: ${{ matrix.os }}
 
     steps:
       - name: Checkout code
         uses: actions/checkout@v2
         with:
           submodules: true
 
-      - name: Retrieve opam cache
-        uses: actions/cache@v2
-        if: runner.os != 'Windows'
-        id: cache-opam
-        with:
-          path: ~/.opam
-          key: v1-${{ runner.os }}-opam-${{ matrix.ocaml-version }}-${{ hashFiles('*.opam.locked') }}
-          restore-keys: |
-            v1-${{ runner.os }}-opam-${{ matrix.ocaml-version }}-
-
       - name: Use latest LTS Node.js release
-        uses: actions/setup-node@v1
+        uses: actions/setup-node@v2
         with:
           node-version: 14.x
 
-      - name: Use OCaml ${{ matrix.ocaml-version }}
-        uses: avsm/setup-ocaml@v1
-        with:
-          ocaml-version: ${{ matrix.ocaml-version }}
-
-      - name: Print OCaml version
-        run: opam exec -- ocaml --version
-
-      - name: Pin local packages
-        run: |
-          opam pin jsonrpc.dev . --no-action
-          opam pin lsp.dev . --no-action
-          opam pin ocaml-lsp-server.dev . --no-action
-
-      - name: Install OCaml secondary compiler
-        if: steps.cache-opam.outputs.cache-hit != 'true' && runner.os != 'Windows' && (matrix.ocaml-version == '4.06.1')
-        run: opam install ocamlfind-secondary
+      - run: yarn --frozen-lockfile
+        working-directory: ocaml-lsp-server/test/e2e
 
-      - name: Install Opam dependencies
-        if: steps.cache-opam.outputs.cache-hit != 'true'
-        run: |
-          opam install . --with-doc --with-test --locked --ignore-constraints-on=ocaml
-          opam install ocamlformat.0.17.0 --skip-updates
+      - name: Use OCaml ${{ matrix.ocaml-compiler }}
+        uses: avsm/setup-ocaml@v2
+        with:
+          ocaml-compiler: ${{ matrix.ocaml-compiler }}
+          dune-cache: ${{ matrix.os != 'macos-latest' }}
+          opam-depext-flags: --with-test
 
-      - name: Fix broken opam state
-        if: steps.cache-opam.outputs.cache-hit == 'true'
-        run: opam upgrade --fixup
+      - run: opam depext ocamlformat=$(cat .ocamlformat | grep version | cut -d '=' -f 2) --install
 
-      - name: Install Node dependencies
-        working-directory: ocaml-lsp-server/test/e2e
-        run: |
-          yarn --frozen-lockfile
+      - run: opam install . --with-test
 
-      - name: Build
-        run: opam exec -- make all
+      - run: opam exec -- make all
 
-      - name: Check formatting
-        if: runner.os == 'Linux'
-        run: opam exec -- make fmt
+      - run: opam exec -- make test-ocaml
+        timeout-minutes: 3
 
-      - name: Run OCaml tests
-        run: opam exec -- make test-ocaml
+      - run: opam exec -- make test-e2e
+        timeout-minutes: 3
```

Comment:
Set a timeout because of test instability on Windows and automatically fail the job if the issue occurs.